### ISSUE TITLE
Truncated output in Journald since v244 when producing output in sub-processes

### ISSUE THREAD

**Issue by cpaelzer** [systemd/systemd#15654]
**systemd version the issue has been seen with**
Since 244.1

I was comparing the Ubuntu versions that were built for 20.04 in the past and got:

- 245.4-4ubuntu3 bad
- 245.2-1ubuntu1 bad
- 244.3-1ubuntu1 bad
- 244.1-0ubuntu3 bad
- 243-3ubuntu1 good
- 242-7ubuntu3 good

**Used distribution**
Ubuntu 20.04
But it was now also reported to trigger on recent Fedora.

**Expected behaviour you didn't see**
Output like

> Apr 30 18:56:40 f systemd[1]: Stopping Test Truncate...
> Apr 30 18:56:40 f test.sh[1165]: This will
> Apr 30 18:56:40 f test.sh[1167]: usually fail
> Apr 30 18:56:40 f test.sh[1168]: and be truncate
> Apr 30 18:56:40 f systemd[1]: test.service: Succeeded.
> Apr 30 18:56:40 f systemd[1]: Stopped Test Truncate.
> 

**Unexpected behaviour you saw**

> Apr 30 18:56:40 f systemd[1]: Stopping Test Truncate...
> Apr 30 18:56:40 f test.sh[1165]: T
> Apr 30 18:56:40 f test.sh[1167]: T
> Apr 30 18:56:40 f test.sh[1167]: sually fai
> Apr 30 18:56:40 f test.sh[1165]: s
> Apr 30 18:56:40 f test.sh[1168]: s
> Apr 30 18:56:40 f test.sh[1168]: nd be truncate
> Apr 30 18:56:40 f test.sh[1165]: n
> Apr 30 18:56:40 f systemd[1]: test.service: Succeeded.
> Apr 30 18:56:40 f systemd[1]: Stopped Test Truncate.

**Steps to reproduce the problem**
Use a test service like /etc/systemd/system/test.service:
```
[Unit]
Description=Test Truncate
After=network.target

[Service]
ExecStart=/usr/lib/test.sh long-test-for-start
ExecStop=/usr/lib/test.sh long-test-for-stop
Type=oneshot
RemainAfterExit=yes
StandardOutput=journal+console
TimeoutStopSec=0

[Install]
WantedBy=multi-user.target
```

And a test script like /usr/lib/test.sh:
```
#!/bin/sh
gettext "This will"
echo
gettext "usually fail"
echo
gettext "and be truncated"
echo
```

Note, this is simplified for testing - in real life this affects libvirts helper to shutdown guests on host shutdown - see [LP 1875708](https://bugs.launchpad.net/ubuntu/+source/libvirt/+bug/1875708).

I [reported this to libvirt trying to fix the behavior](https://www.redhat.com/archives/libvir-list/2020-April/msg01474.html) by ensuring the subprograms do NOT emit to stdout directly. This works, but we don't understand why and wondered if it might be a bug in journald instead.

Daniel did some [debugging on this](https://www.redhat.com/archives/libvir-list/2020-April/msg01527.html) and found:

> I strace'd systemd-journald and the data it receives looks fine,
> but the data systemd-journald writes into the journal, however, is
> completely fubar:

```
recvmsg(32, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="Resuming guest core1: ", iov_len=2054}], msg_iovlen=1, msg_control=[{cmsg_len=28, cmsg_level=SOL_SOCKET, cmsg_type=SCM_CREDENTIALS, cmsg_data={pid=70624, uid=0, gid=0}}], msg_controllen=32, msg_flags=MSG_CTRUNC|MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 22
writev(33, [{iov_base="[ 2741.146550] ", iov_len=15}, {iov_base="libvirt-guests.sh", iov_len=17}, {iov_base="[70605]: ", iov_len=9}, {iov_base="R", iov_len=1}, {iov_base="\n", iov_len=1}], 5) = 43
writev(33, [{iov_base="[ 2741.163787] ", iov_len=15}, {iov_base="libvirt-guests.sh", iov_len=17}, {iov_base="[70624]: ", iov_len=9}, {iov_base="R", iov_len=1}, {iov_base="\n", iov_len=1}], 5) = 43
writev(33, [{iov_base="[ 2742.518626] ", iov_len=15}, {iov_base="libvirt-guests.sh", iov_len=17}, {iov_base="[70624]: ", iov_len=9}, {iov_base="esuming guest core1:", iov_len=20}, {iov_base="\n", iov_len=1}], 5) = 62
writev(33, [{iov_base="[ 2742.531644] ", iov_len=15}, {iov_base="libvirt-guests.sh", iov_len=17}, {iov_base="[70655]: ", iov_len=9}, {iov_base="esum", iov_len=4}, {iov_base="\n", iov_len=1}], 5) = 46
recvmsg(32, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="Resuming guest core2: ", iov_len=2054}], msg_iovlen=1, msg_control=[{cmsg_len=28, cmsg_level=SOL_SOCKET, cmsg_type=SCM_CREDENTIALS, cmsg_data={pid=70662, uid=0, gid=0}}], msg_controllen=32, msg_flags=MSG_CTRUNC|MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 22

writev(33, [{iov_base="[ 2742.687436] ", iov_len=15}, {iov_base="libvirt-guests.sh", iov_len=17}, {iov_base="[70605]: ", iov_len=9}, {iov_base="e", iov_len=1}, {iov_base="\n", iov_len=1}], 5) = 43
writev(33, [{iov_base="[ 2742.690716] ", iov_len=15}, {iov_base="libvirt-guests.sh", iov_len=17}, {iov_base="[70662]: ", iov_len=9}, {iov_base="e", iov_len=1}, {iov_base="\n", iov_len=1}], 5) = 43
writev(33, [{iov_base="[ 2744.015012] ", iov_len=15}, {iov_base="libvirt-guests.sh", iov_len=17}, {iov_base="[70662]: ", iov_len=9}, {iov_base="esuming guest core2:", iov_len=20}, {iov_base="\n", iov_len=1}], 5) = 62
writev(33, [{iov_base="[ 2744.019848] ", iov_len=15}, {iov_base="libvirt-guests.sh", iov_len=17}, {iov_base="[70694]: ", iov_len=9}, {iov_base="esum", iov_len=4}, {iov_base="\n", iov_len=1}], 5) = 46
writev(33, [{iov_base="[ 2744.036819] ", iov_len=15}, {iov_base="libvirt-guests.sh", iov_len=17}, {iov_base="[70605]: ", iov_len=9}, {iov_base="e", iov_len=1}, {iov_base="\n", iov_len=1}], 5) = 43
```

Does that ring any bell of related systemd changes or known issues?

**Comment by benjarobin** [comment]
The commit that introduces this regression is 09d0b46ab61bebafe5bdc1be95ee153dfb13d6bc (#13771)
```
$ git bisect start v245 v243
Bissection : 1044 révisions à tester après ceci (à peu près 10 étapes)
[3d85f19cedf1f7dcfc06df4b88b626e3305c0808] Merge pull request #14305 from ssahani/tc-fq
$ git bisect bad
Bissection : 521 révisions à tester après ceci (à peu près 9 étapes)
[8cb28504a5edb365942e8ee8a957374d46b1384f] tree-wide: drop input.h when missing_input.h is included
$ git bisect good
Bissection : 260 révisions à tester après ceci (à peu près 8 étapes)
[88a36d369015674e3f9eb20f0fc4d33beedd91bb] varlink: port varlink code over to use getdtablesize() for sizing number of concurrent connections
$ git bisect bad
Bissection : 130 révisions à tester après ceci (à peu près 7 étapes)
[e013e10d0e96917b627d54f20e3dba80a17df0d5] ask-password: don't hit assert() when we query pw which the user C-d and caching is enabled
git bisect bad
Bissection : 64 révisions à tester après ceci (à peu près 6 étapes)
[1f6597a84cd3891a69a630c6c4b168da46b1c160] man: mention $RUNTIME_DIRECTORY & friends in environment list
$ git bisect bad
Bissection : 32 révisions à tester après ceci (à peu près 5 étapes)
[3e1db806b0c18fd6138886ce67fac2655f09caef] core: change top-level drop-in from -.service.d to service.d
git bisect bad
Bissection : 16 révisions à tester après ceci (à peu près 4 étapes)
[69d7eba1880095f4a9bf9350de777fe7d370c188] sd-radv: fix memleak
git bisect good
Bissection : 8 révisions à tester après ceci (à peu près 3 étapes)
[9087384d392557fdc7cc4a3b07ff3b9b45d11524] Merge pull request #13944 from yuwata/network-split-struct-prefix
$ git bisect bad
Bissection : 3 révisions à tester après ceci (à peu près 2 étapes)
[09d0b46ab61bebafe5bdc1be95ee153dfb13d6bc] journal: refresh cached credentials of stdout streams
$ git bisect bad
Bissection : 1 révision à tester après ceci (à peu près 1 étape)
[d36b573ef7a5a872cc5a4079cb1961e0d066a7b2] Merge pull request #13935 from poettering/bootctl-random-seed-mkdir
$ git bisect good
Bissection : 0 révision à tester après ceci (à peu près 0 étape)
[d7d31692bf7cde5dce7f4ed3cae429a5b302a9f0] hwdb: add XKB_FIXED_MODEL to the keyboard hwdb
benjarobin ~/Application/systemd-git (git)-[v244-rc1~97|bisect] $ git bisect good
09d0b46ab61bebafe5bdc1be95ee153dfb13d6bc is the first bad commit
commit 09d0b46ab61bebafe5bdc1be95ee153dfb13d6bc
Author: Lorenz Bauer <lmb@cloudflare.com>
Date:   Mon Nov 4 16:35:46 2019 +0000

    journal: refresh cached credentials of stdout streams

    journald assumes that getsockopt(SO_PEERCRED) correctly identifies the
    process on the remote end of the socket. However, this is incorrect
    according to man 7 socket:

        The returned  credentials  are  those that were in effect at the
        time of the call to connect(2) or socketpair(2).

    This becomes a problem when a new process inherits the stdout stream
    from a parent. First, log messages from the child process will
    be attributed to the parent. Second, the struct ucred used by journald
    becomes invalid as soon as the parent exits. Further sendmsg calls then
    fail with ENOENT. Logs for the child process then vanish from the journal.

    Fix this by using recvmsg on the stdout stream, and refreshing the cached
    struct ucred if SCM_CREDENTIALS indicate a new process.

    Fixes #13708

 src/journal/journald-stream.c        | 49 ++++++++++++++++++++++++++++++++++--
 test/TEST-04-JOURNAL/test-journal.sh | 16 ++++++++++++
 2 files changed, 63 insertions(+), 2 deletions(-)
```

**Comment by benjarobin** [comment]
I am working on a patch, I think I found the issue...

**Comment by berrange** [comment]
I would note that trying to associate  PID with individual data buffers read is really unhelpful when the unit is implemented as a shell script. In the libvirt-guests.service case, we have a shell script and that uses the "gettext" helper command for translation. As a result each piece of text is associated with a new pid as each new piece of text is printed by a new invocation of the "gettext" helper. IOW, the original behaviour of logging the PID of the primary process that first created the socket was more desirable.

**Comment by benjarobin** [comment]
I did not changed that, this behavior was introduced in order to fix #13708. This MR only fix some "edge" cases...
